### ISSUE TITLE
About wait/waitpid syscall[BUG]

### ISSUE THREAD

**Issue by Antone077** [occlum/occlum#1245]
# Describe the bug
Hello developer! 
      Recently I'm writing some syscall tests to detect Occlum, I want to make sense the support level of Occlum in syscall aspect. Now I use wait or waitpid syscall to wait for process to change state. And I find when I use vfork syscall to create a child process, then I use _exit(0) syscall which can exit this child process. Finally, I use wait or waitpid syscall which can record the state when exiting. But I find the difference between the Occlum and my host machine. It will report error when I run my test in Occlum. The error message is following:
     `wait error: No child process`

# Expected behavior
      Expected behavior is that wait syscall can record the status, and then I can check the status message.

# Environment

- HW: [SGX1] 
- OS: [Ubuntu20.04]
- Occlum version: [e.g. 0.29.3]

# Additional context
        By the way, this test can run in 0.29.3. The newer Occlum version such as 0.29.4/0.29.5, I cann't run any C language tests.
It will report error like this:
`[ERROR] occlum-pal: Failed to create enclave with error code 0x1: Unexpected error occurred. (line 152, file src/pal_enclave.c)`
        I am not sure whether my computer settings have some wrong?But I use 0.29.3 version ,I can run successfully with the same steps.

**Comment by jessehui** [comment]
Thanks for using Occlum. 

> And I find when I use vfork syscall to create a child process, then I use _exit(0) syscall which can exit this child process. Finally, I use wait or waitpid syscall which can record the state when exiting. But I find the difference between the Occlum and my host machine. It will report error when I run my test in Occlum. The error message is following:
wait error: No child process

The reason is that in Occlum's vfork implementation, there is no actual child process at all. Child process created with vfork is supposed to do nothing but `exit` or `execve` because the vfork-ed process is reusing the parent's address space. And the child pid returned by vfork is not actually corresponding to a real process. Thus wait will not work. This is a limitation of the current implementation. I will take a look to see if it can be fixed. Please provide us with a minimum reproducible version of your test case. Thank you.

**Comment by jessehui** [comment]
#1247 This PR should fix your problem.

**Comment by Antone077** [comment]
@jessehui Thank you for your answer and fix this issue, I see the code has been submitted successfully. Does it mean that I can use the latest version to verify this issue?

**Comment by Antone077** [comment]
@jessehui
Another problem is that when I run the latest version of Occlum, some weird errors occur. Here are the steps I run:
step1:`docker run -it --privileged -v /dev/sgx:/dev/sgx occlum/occlum:0.29.5-ubuntu20.04`
step2: `cd demo/hello_c && make`

Then I want to run hello_world on Occlum
`mkdir occlum_workspace && cd occlum_workspace`
`occlum init && rm -rf image`
`copy_bom -f ../hello.yaml --root image --include-dir /opt/occlum/etc/template`
`occlum build`
`occlum run /bin/hello_world`

The result is that a strange error is returned, not what I expected：
`[ERROR] occlum-pal: Failed to create enclave with error code 0x1: Unexpected error occurred. (line 152, file src/pal_enclave.c)`

But I use the above steps, it works perfectly on the 0.29.3 version.

**Comment by Antone077** [comment]
> Thanks for using Occlum.
> 
> > And I find when I use vfork syscall to create a child process, then I use _exit(0) syscall which can exit this child process. Finally, I use wait or waitpid syscall which can record the state when exiting. But I find the difference between the Occlum and my host machine. It will report error when I run my test in Occlum. The error message is following:
> > wait error: No child process
> 
> The reason is that in Occlum's vfork implementation, there is no actual child process at all. Child process created with vfork is supposed to do nothing but `exit` or `execve` because the vfork-ed process is reusing the parent's address space. And the child pid returned by vfork is not actually corresponding to a real process. Thus wait will not work. This is a limitation of the current implementation. I will take a look to see if it can be fixed. Please provide us with a minimum reproducible version of your test case. Thank you.

Here is my test case.

```
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>

int main() {
    pid_t pid, child_pid;
    int status;
    pid = vfork(); 
    if (pid < 0) {
        fprintf(stderr, "Failed to fork a new process\n");
        exit(EXIT_FAILURE);
    } else if (pid == 0) {
        printf("Child process is running\n");
        exit(EXIT_SUCCESS);
    } else {
        printf("Parent process is waiting for the child\n");
        child_pid = waitpid(pid, &status, 0); 
        if (child_pid < 0) {  
            fprintf(stderr, "Failed to wait for the child process\n");
            exit(EXIT_FAILURE);
        }
        if (WIFEXITED(status)) { 
            printf("Child process exited with status %d\n", WEXITSTATUS(status));
        } else {
            printf("Child process terminated abnormally\n");
        }
    }
    return 0;
}
```
And this test case will report error, because of the child_pid < 0.

**Comment by jessehui** [comment]
> Does it mean that I can use the latest version to verify this issue?

No, not yet. You need to patch the code and install Occlum by yourself.

> Another problem is that when I run the latest version of Occlum, some weird errors occur. Here are the steps I run:
step1:docker run -it --privileged -v /dev/sgx:/dev/sgx occlum/occlum:0.29.5-ubuntu20.04
step2: cd demo/hello_c && make
 Then I want to run hello_world on Occlum
mkdir occlum_workspace && cd occlum_workspace
occlum init && rm -rf image
copy_bom -f ../hello.yaml --root image --include-dir /opt/occlum/etc/template
occlum build
occlum run /bin/hello_world
The result is that a strange error is returned, not what I expected：
[ERROR] occlum-pal: Failed to create enclave with error code 0x1: Unexpected error occurred. (line 152, file src/pal_enclave.c)
But I use the above steps, it works perfectly on the 0.29.3 version.

This is weird. Could you list the platform and the OS version you are using? Also please try the Intel SGX SDK sample and see if it works. If it can't run either, I think it is related to the platform or driver. If it can run, please copy the "Enclave.xml" file from the `occlum_workspace/build/Enclave.xml` and use it in the SDK sample and see if it works this time. It is then probably due to some new configuration we added after 0.29.3 that your platform doesn't support.

**Comment by Antone077** [comment]
> > Does it mean that I can use the latest version to verify this issue?
> 
> No, not yet. You need to patch the code and install Occlum by yourself.
> 
> > Another problem is that when I run the latest version of Occlum, some weird errors occur. Here are the steps I run:
> > step1:docker run -it --privileged -v /dev/sgx:/dev/sgx occlum/occlum:0.29.5-ubuntu20.04
> > step2: cd demo/hello_c && make
> > Then I want to run hello_world on Occlum
> > mkdir occlum_workspace && cd occlum_workspace
> > occlum init && rm -rf image
> > copy_bom -f ../hello.yaml --root image --include-dir /opt/occlum/etc/template
> > occlum build
> > occlum run /bin/hello_world
> > The result is that a strange error is returned, not what I expected：
> > [ERROR] occlum-pal: Failed to create enclave with error code 0x1: Unexpected error occurred. (line 152, file src/pal_enclave.c)
> > But I use the above steps, it works perfectly on the 0.29.3 version.
> 
> This is weird. Could you list the platform and the OS version you are using? Also please try the Intel SGX SDK sample and see if it works. If it can't run either, I think it is related to the platform or driver. If it can run, please copy the "Enclave.xml" file from the `occlum_workspace/build/Enclave.xml` and use it in the SDK sample and see if it works this time. It is then probably due to some new configuration we added after 0.29.3 that your platform doesn't support.

Thank you for your reply, I‘m using Ubuntu 22.04.2 TLS and my kernel version is 5.19.0-40-generic. I will try the way you provided，Thank you again for your response.

**Comment by Antone077** [comment]
> > Does it mean that I can use the latest version to verify this issue?
> 
> No, not yet. You need to patch the code and install Occlum by yourself.
> 
> > Another problem is that when I run the latest version of Occlum, some weird errors occur. Here are the steps I run:
> > step1:docker run -it --privileged -v /dev/sgx:/dev/sgx occlum/occlum:0.29.5-ubuntu20.04
> > step2: cd demo/hello_c && make
> > Then I want to run hello_world on Occlum
> > mkdir occlum_workspace && cd occlum_workspace
> > occlum init && rm -rf image
> > copy_bom -f ../hello.yaml --root image --include-dir /opt/occlum/etc/template
> > occlum build
> > occlum run /bin/hello_world
> > The result is that a strange error is returned, not what I expected：
> > [ERROR] occlum-pal: Failed to create enclave with error code 0x1: Unexpected error occurred. (line 152, file src/pal_enclave.c)
> > But I use the above steps, it works perfectly on the 0.29.3 version.
> 
> This is weird. Could you list the platform and the OS version you are using? Also please try the Intel SGX SDK sample and see if it works. If it can't run either, I think it is related to the platform or driver. If it can run, please copy the "Enclave.xml" file from the `occlum_workspace/build/Enclave.xml` and use it in the SDK sample and see if it works this time. It is then probably due to some new configuration we added after 0.29.3 that your platform doesn't support.

Is it only on machines that support SGX2, and can run successfully occlum version:0.29.5?

**Comment by volcano0dr** [comment]
> Is it only on machines that support SGX2, and can run successfully occlum version:0.29.5?

I will test occlum 0.29.5 on SGX1 hardware platform.

**Comment by Antone077** [comment]
> > Is it only on machines that support SGX2, and can run successfully occlum version:0.29.5?
> 
> I will test occlum 0.29.5 on SGX1 hardware platform.

I am happy to get your reply and help because I only have SGX1 hardware platform, so I can't test for myself.
